### PR TITLE
Find Homebrew OpenSSL on M1.

### LINKED ISSUE #426: Logic for dealing with openssl on macOS needs to be updated
in rustls-mio/tests/common/mod.rs:
```
pub fn openssl_find() -> String {
    if let Ok(dir) = env::var("OPENSSL_DIR") {
        return format!("{}/bin/openssl", dir);
    }

    // We need a homebrew openssl, because OSX comes with
    // 0.9.8y or something equally ancient!
    if cfg!(target_os = "macos") {
        if Path::new("/usr/local/opt/openssl@1.1/bin/openssl").is_file() {
            return "/usr/local/opt/openssl@1.1/bin/openssl".to_string();
        } else {
            return "/usr/local/opt/openssl/bin/openssl".to_string();
        }
    }

    "openssl".to_string()
}
```

In my M1 Mac with Big Sur, `openssl version` reports "LibreSSL 2.8.3" and `which openssl` reports "/usr/bin/openssl." I hope that the x86_64 version of macOS is similar now? If so, maybe we can get rid of the homebrew stuff?

I used `OPENSSL_DIR=$(dirname $(dirname $(which openssl))) cargo test` (i.e. `OPENSSL_DIR=/usr`) and test `no_ecdhe` failed with:

<details>
```
We expected to find 'TLS error: AlertReceived\(HandshakeFailure\)' in the following output:
Output { status: ExitStatus(ExitStatus(0)), stdout: "HTTP/1.0 200 ok\r\nContent-type: text/html\r\n\r\n<HTML><BODY BGCOLOR=\"#ffffff\">\n<pre>\n\ns_server -accept 8010 -key /var/folders/wr/tqrw8vl54h7d85hfksc_g0dh0000gn/T/rustls-a1de3e9e/rsa/end.key -cert /var/folders/wr/tqrw8vl54h7d85hfksc_g0dh0000gn/T/rustls-a1de3e9e/rsa/end.cert -key2 /var/folders/wr/tqrw8vl54h7d85hfksc_g0dh0000gn/T/rustls-a1de3e9e/rsa/end.key -cert2 /var/folders/wr/tqrw8vl54h7d85hfksc_g0dh0000gn/T/rustls-a1de3e9e/rsa/end.cert -CAfile /var/folders/wr/tqrw8vl54h7d85hfksc_g0dh0000gn/T/rustls-a1de3e9e/rsa/end.chain -no_ecdhe -www \nSecure Renegotiation IS supported\nCiphers supported in s_server binary\nTLSv1/SSLv3:ECDHE-ECDSA-CHACHA20-POLY1305TLSv1/SSLv3:ECDHE-RSA-CHACHA20-POLY1305\nTLSv1/SSLv3:DHE-RSA-CHACHA20-POLY1305TLSv1/SSLv3:ECDHE-RSA-AES256-GCM-SHA384\nTLSv1/SSLv3:ECDHE-ECDSA-AES256-GCM-SHA384TLSv1/SSLv3:ECDHE-RSA-AES256-SHA384  \nTLSv1/SSLv3:ECDHE-ECDSA-AES256-SHA384TLSv1/SSLv3:ECDHE-RSA-AES256-SHA     \nTLSv1/SSLv3:ECDHE-ECDSA-AES256-SHA   TLSv1/SSLv3:DHE-RSA-AES256-GCM-SHA384\nTLSv1/SSLv3:DHE-RSA-AES256-SHA256    TLSv1/SSLv3:DHE-RSA-AES256-SHA       \nTLSv1/SSLv3:GOST2012256-GOST89-GOST89TLSv1/SSLv3:DHE-RSA-CAMELLIA256-SHA256\nTLSv1/SSLv3:DHE-RSA-CAMELLIA256-SHA  TLSv1/SSLv3:GOST2001-GOST89-GOST89   \nTLSv1/SSLv3:AES256-GCM-SHA384        TLSv1/SSLv3:AES256-SHA256            \nTLSv1/SSLv3:AES256-SHA               TLSv1/SSLv3:CAMELLIA256-SHA256       \nTLSv1/SSLv3:CAMELLIA256-SHA          TLSv1/SSLv3:ECDHE-RSA-AES128-GCM-SHA256\nTLSv1/SSLv3:ECDHE-ECDSA-AES128-GCM-SHA256TLSv1/SSLv3:ECDHE-RSA-AES128-SHA256  \nTLSv1/SSLv3:ECDHE-ECDSA-AES128-SHA256TLSv1/SSLv3:ECDHE-RSA-AES128-SHA     \nTLSv1/SSLv3:ECDHE-ECDSA-AES128-SHA   TLSv1/SSLv3:DHE-RSA-AES128-GCM-SHA256\nTLSv1/SSLv3:DHE-RSA-AES128-SHA256    TLSv1/SSLv3:DHE-RSA-AES128-SHA       \nTLSv1/SSLv3:DHE-RSA-CAMELLIA128-SHA256TLSv1/SSLv3:DHE-RSA-CAMELLIA128-SHA  \nTLSv1/SSLv3:AES128-GCM-SHA256        TLSv1/SSLv3:AES128-SHA256            \nTLSv1/SSLv3:AES128-SHA               TLSv1/SSLv3:CAMELLIA128-SHA256       \nTLSv1/SSLv3:CAMELLIA128-SHA          TLSv1/SSLv3:ECDHE-RSA-RC4-SHA        \nTLSv1/SSLv3:ECDHE-ECDSA-RC4-SHA      TLSv1/SSLv3:RC4-SHA                  \nTLSv1/SSLv3:RC4-MD5                  TLSv1/SSLv3:ECDHE-RSA-DES-CBC3-SHA   \nTLSv1/SSLv3:ECDHE-ECDSA-DES-CBC3-SHA TLSv1/SSLv3:EDH-RSA-DES-CBC3-SHA     \nTLSv1/SSLv3:DES-CBC3-SHA             \n---\nCiphers common between both SSL end points:\nECDHE-ECDSA-CHACHA20-POLY1305 ECDHE-RSA-CHACHA20-POLY1305 ECDHE-ECDSA-AES256-GCM-SHA384\nECDHE-ECDSA-AES128-GCM-SHA256 ECDHE-RSA-AES256-GCM-SHA384 ECDHE-RSA-AES128-GCM-SHA256\n---\nNew, TLSv1/SSLv3, Cipher is ECDHE-RSA-CHACHA20-POLY1305\nSSL-Session:\n    Protocol  : TLSv1.2\n    Cipher    : ECDHE-RSA-CHACHA20-POLY1305\n    Session-ID: \n    Session-ID-ctx: 01000000\n    Master-Key: F146672FEF039EEC54EA595FBA1D6E9ECD3FEA574083EC69664B111403DDAE9FC05C7F1F87EB1E12BAEA4DEBE1D73A2D\n    Start Time: 1606888176\n    Timeout   : 7200 (sec)\n    Verify return code: 0 (ok)\n---\n   0 items in the session cache\n   0 client connects (SSL_connect())\n   0 client renegotiates (SSL_connect())\n   0 client connects that finished\n   2 server accepts (SSL_accept())\n   0 server renegotiates (SSL_accept())\n   1 server accepts that finished\n   0 session cache hits\n   1 session cache misses\n   0 session cache timeouts\n   0 callback cache hits\n   0 cache full overflows (128 allowed)\n---\nno client certificate available\n</BODY></HTML>\r\n\r\nEOF\nConnection closed\n", stderr: "[2020-12-02T05:49:36Z DEBUG rustls::anchors] add_pem_file processed 1 valid and 0 invalid certs\n[2020-12-02T05:49:36Z DEBUG rustls::client::hs] No cached session for DNSNameRef(\"localhost\")\n[2020-12-02T05:49:36Z DEBUG rustls::client::hs] Not resuming any session\n[2020-12-02T05:49:36Z TRACE rustls::client::hs] Sending ClientHello Message {\n        typ: Handshake,\n        version: TLSv1_0,\n        payload: Handshake(\n            HandshakeMessagePayload {\n                typ: ClientHello,\n                payload: ClientHello(\n                    ClientHelloPayload {\n                        client_version: TLSv1_2,\n                        random: Random(\n                            [\n                                244,\n                                1,\n                                80,\n                                245,\n                                77,\n                                176,\n                                151,\n                                13,\n                                176,\n                                25,\n                                197,\n                                156,\n                                152,\n                                169,\n                                35,\n                                228,\n                                139,\n                                154,\n                                8,\n                                250,\n                                174,\n                                16,\n                                159,\n                                150,\n                                247,\n                                127,\n                                237,\n                                44,\n                                167,\n                                198,\n                                148,\n                                183,\n                            ],\n                        ),\n                        session_id: SessionID(\n                            140,\n                            134,\n                            48,\n                            51,\n                            179,\n                            180,\n                            94,\n                            32,\n                            111,\n                            84,\n                            93,\n                            63,\n                            129,\n                            209,\n                            163,\n                            62,\n                            204,\n                            195,\n                            189,\n                            55,\n                            17,\n                            3,\n                            194,\n                            228,\n                            208,\n                            44,\n                            174,\n                            148,\n                            189,\n                            102,\n                            84,\n                            116,\n                        ),\n                        cipher_suites: [\n                            TLS13_CHACHA20_POLY1305_SHA256,\n                            TLS13_AES_256_GCM_SHA384,\n                            TLS13_AES_128_GCM_SHA256,\n                            TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256,\n                            TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256,\n                            TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,\n                            TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,\n                            TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,\n                            TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,\n                            TLS_EMPTY_RENEGOTIATION_INFO_SCSV,\n                        ],\n                        compression_methods: [\n                            Null,\n                        ],\n                        extensions: [\n                            SupportedVersions(\n                                [\n                                    TLSv1_3,\n                                    TLSv1_2,\n                                ],\n                            ),\n                            ServerName(\n                                [\n                                    ServerName {\n                                        typ: HostName,\n                                        payload: HostName(\n                                            DNSName(\n                                                \"localhost\",\n                                            ),\n                                        ),\n                                    },\n                                ],\n                            ),\n                            ECPointFormats(\n                                [\n                                    Uncompressed,\n                                ],\n                            ),\n                            NamedGroups(\n                                [\n                                    X25519,\n                                    secp384r1,\n                                    secp256r1,\n                                ],\n                            ),\n                            SignatureAlgorithms(\n                                [\n                                    ECDSA_NISTP384_SHA384,\n                                    ECDSA_NISTP256_SHA256,\n                                    ED25519,\n                                    RSA_PSS_SHA512,\n                                    RSA_PSS_SHA384,\n                                    RSA_PSS_SHA256,\n                                    RSA_PKCS1_SHA512,\n                                    RSA_PKCS1_SHA384,\n                                    RSA_PKCS1_SHA256,\n                                ],\n                            ),\n                            ExtendedMasterSecretRequest,\n                            CertificateStatusRequest(\n                                OCSP(\n                                    OCSPCertificateStatusRequest {\n                                        responder_ids: [],\n                                        extensions: PayloadU16(\n                                            [],\n                                        ),\n                                    },\n                                ),\n                            ),\n                            KeyShare(\n                                [\n                                    KeyShareEntry {\n                                        group: X25519,\n                                        payload: PayloadU16(\n                                            [\n                                                213,\n                                                229,\n                                                229,\n                                                247,\n                                                243,\n                                                13,\n                                                215,\n                                                70,\n                                                185,\n                                                132,\n                                                175,\n                                                191,\n                                                204,\n                                                33,\n                                                243,\n                                                248,\n                                                224,\n                                                128,\n                                                243,\n                                                241,\n                                                119,\n                                                213,\n                                                1,\n                                                177,\n                                                251,\n                                                235,\n                                                255,\n                                                46,\n                                                107,\n                                                96,\n                                                50,\n                                                26,\n                                            ],\n                                        ),\n                                    },\n                                ],\n                            ),\n                            PresharedKeyModes(\n                                [\n                                    PSK_DHE_KE,\n                                ],\n                            ),\n                            SessionTicketRequest,\n                        ],\n                    },\n                ),\n            },\n        ),\n    }\n[2020-12-02T05:49:36Z TRACE mio::poll] registering event source with poller: token=Token(0), interests=READABLE | WRITABLE\n[2020-12-02T05:49:36Z TRACE mio::poll] reregistering event source with poller: token=Token(0), interests=READABLE\n[2020-12-02T05:49:36Z TRACE rustls::client::hs] We got ServerHello ServerHelloPayload {\n        legacy_version: TLSv1_2,\n        random: Random(\n            [\n                196,\n                26,\n                248,\n                171,\n                31,\n                219,\n                209,\n                102,\n                114,\n                106,\n                81,\n                92,\n                129,\n                65,\n                157,\n                61,\n                44,\n                50,\n                134,\n                203,\n                219,\n                130,\n                123,\n                41,\n                153,\n                108,\n                182,\n                181,\n                96,\n                19,\n                228,\n                215,\n            ],\n        ),\n        session_id: SessionID,\n        cipher_suite: TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256,\n        compression_method: Null,\n        extensions: [\n            ServerNameAck,\n            RenegotiationInfo(\n                PayloadU8(\n                    [],\n                ),\n            ),\n            ECPointFormats(\n                [\n                    Uncompressed,\n                ],\n            ),\n            SessionTicketAck,\n        ],\n    }\n[2020-12-02T05:49:36Z DEBUG rustls::client::hs] ALPN protocol is None\n[2020-12-02T05:49:36Z DEBUG rustls::client::hs] Using ciphersuite TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256\n[2020-12-02T05:49:36Z DEBUG rustls::client::hs] Server supports tickets\n[2020-12-02T05:49:36Z DEBUG rustls::client::tls12] ECDHE curve is ECParameters { curve_type: NamedCurve, named_group: X25519 }\n[2020-12-02T05:49:36Z TRACE rustls::client::tls12] Server cert is [Certificate([...])]\n[2020-12-02T05:49:36Z DEBUG rustls::client::tls12] Server DNS name is DNSName(\"localhost\")\n[2020-12-02T05:49:36Z TRACE mio::poll] reregistering event source with poller: token=Token(0), interests=READABLE | WRITABLE\n[2020-12-02T05:49:36Z TRACE mio::poll] reregistering event source with poller: token=Token(0), interests=READABLE\n[2020-12-02T05:49:36Z DEBUG rustls::client::tls12] Session saved\n[2020-12-02T05:49:36Z TRACE mio::poll] reregistering event source with poller: token=Token(0), interests=READABLE | WRITABLE\n[2020-12-02T05:49:36Z TRACE mio::poll] reregistering event source with poller: token=Token(0), interests=READABLE\n[2020-12-02T05:49:36Z TRACE mio::poll] reregistering event source with poller: token=Token(0), interests=READABLE\n" }
thread 'no_ecdhe' panicked at 'Test failed', rustls-mio/tests/common/mod.rs:448:17
</details>

### PR DESCRIPTION
**Pull request by sayrer**
Fixes #426 for me on a new M1 Air. I think the built-in LibreSSL could be made to work as well, if that's desirable.

### CODE REVIEW

File: rustls-mio/tests/common/mod.rs [hunk 1]
```diff
@@ -192,13 +192,26 @@ pub fn openssl_find() -> String {
...
-            return "/usr/local/opt/openssl/bin/openssl".to_string();
+        match process::Command::new("brew")
+            .args(&["--prefix", "openssl"])
+            .output() {
+            Ok(output) => {
+                let dir = str::from_utf8(&*output.stdout).unwrap().trim();
+                return format!("{}/bin/openssl", dir);
+            },
+            Err(_) => {
+                if Path::new("/usr/local/opt/openssl@1.1/bin/openssl").is_file() {
```
Comment (djc) [inline rustls-mio/tests/common/mod.rs:206]: Maybe extract the possible paths as a const slice of strs?

**Review by djc — COMMENTED** [review]
Looks like a nice improvement!

File: rustls-mio/tests/common/mod.rs (hunk 1 above)
Comment (sayrer) [reply rustls-mio/tests/common/mod.rs]: Sure thing. I try not to make stuff like this too concise, so it's easy to see what happens when something goes wrong, but I don't feel strongly about this case.

**Comment by djc** [comment]
Thanks!

**Pull request merged**